### PR TITLE
Refactor grid method

### PR DESCRIPTION
**Pull request by TomMelt**
The `decomp` tool was due a refactor to simplify the code base.

### CODE REVIEW

File: DomainUtils.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * @brief Compute the overlap between two domains. Will return zero if domains
+ * do not overlap.
+ *
+ * Note that this doesn't check if two domains are neighbours, just that they
+ * overlap in the x or y direction.
+ *
+ * @param d1 First Domain
+ * @param d2 Second Domain
+ */
+int domainOverlap(const Domain d1, const Domain d2, const char dir);
```
Comment (jwallwork23) [inline DomainUtils.hpp:45]: You should add `@param dir Direction in which to compute overlap`, or similar.

File: Grid.hpp [hunk 2]
```diff
@@ -208,39 +195,43 @@ class LIB_EXPORT Grid {
...
     int _num_objects = 0; // Number of grid points ignoring land mask
-    int _num_blks = 0; // Number of grid blocks ignoring land mask
     int _num_nonzero_objects = 0; // Number of non-land grid points
-    int _num_nonzero_blks = 0; // Number of non-land grid blocks
-    std::vector<int> _land_mask = {}; // Land mask values of grid bGpoints
-    std::vector<int> _land_mask_blk = {}; // Land mask values of grid blocks
-    std::vector<int> _sparse_to_dense = {}; // Map from sparse to dense index of blocks
-    std::vector<int> _object_id = {}; // Unique non-land grid block IDs
+    const std::string& _dim0_name; // dim0 name
+    const std::string& _dim1_name; // dim1 name
```
Comment (jwallwork23) [inline Grid.hpp:231]: The comments here aren't particularly helpful.

File: DomainUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#include <cstdio>
+#include <iostream>
+
+int domainOverlap(const Domain d1, const Domain d2, const char dir)
+{
+    int overlap = 0;
+    if (dir == 'x') {
+        overlap = std::min(d1.p2.x, d2.p2.x) - std::max(d1.p1.x, d2.p1.x);
+    } else if (dir == 'y') {
+        overlap = std::min(d1.p2.y, d2.p2.y) - std::max(d1.p1.y, d2.p1.y);
```
Comment (jwallwork23) [inline DomainUtils.cpp:19]: It took me a little while to figure out how this worked but, after drawing some diagrams to convince myself, seems like the right thing to do :)

File: Grid.cpp [hunk 4]
```diff
@@ -168,103 +130,105 @@ Grid::Grid(MPI_Comm comm, const std::string& filename, const std::string& dim0_n
...
+    _local_ext_1 = ceil((float)_global_ext_1 / (_num_procs_1));
+    _global_0 = (_rank / _num_procs_1) * _local_ext_0;
+    _global_1 = (_rank % _num_procs_1) * _local_ext_1;
+
+    if ((_rank / _num_procs_1) == _num_procs_0 - 1) {
+        _local_ext_0 = _global_ext_0 - (_rank / _num_procs_1) * _local_ext_0;
+    }
+    if ((_rank % _num_procs_1) == _num_procs_1 - 1) {
+        _local_ext_1 = _global_ext_1 - (_rank % _num_procs_1) * _local_ext_1;
+    }
```
Comment (jwallwork23) [inline Grid.cpp:168]: Since you define `_global_0` and `_global_1`, you could use them inside the `if` statements rather than recomputing.

**Review by jwallwork23 — APPROVED** [review]
Thanks @TomMelt, some more nice refactoring. Some minor comments as suggestions.

File: DomainUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * @file DomainUtils.cpp
+ *
+ * @date Oct 07, 2024
+ * @author Tom Meltzer <tdm39@cam.ac.uk>
+ */
+
+#include "DomainUtils.hpp"
+#include <algorithm>
+#include <cstdio>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (TomMelt) [reply DomainUtils.cpp:11]: sorry my LSP in my text editor automatically adds includes... but it doesn't automatically remove them when they are no longer needed. I guess I used a `printf` statement and then deleted it :/ 

Good spot :mag:

[the remark this answers is not in the document]
Comment (jwallwork23) [reply DomainUtils.cpp:11]: Ooh fancy!

File: Grid.hpp (hunk 2 above)
Comment (TomMelt) [reply Grid.hpp:231]: agreed but they are wiped in #58. So I won't bother changing them here.

File: DomainUtils.cpp (hunk 3 above)
Comment (TomMelt) [reply DomainUtils.cpp:18]: I had to do the same. So many rectangles drawn ...

**Comment by jwallwork23** [comment]
Thanks for your work on this @TomMelt! Happy for it to be merged.

**Pull request merged**